pyomo/contrib/cspline_external/src/functions.cpp

```diff
@@ -0,0 +1,171 @@
+*/
+#include<fstream>
+#include<unordered_map>
+#include<vector>
+#include<string>
+#include<algorithm>
+
+#include "funcadd.h"
+
+// Don't want to have the overhead of reading parameter files
+// each time, so keep params as long as library remains loaded
+
+// Number of loaded curves
+unsigned int n_cspline = 0;
+
+// Parameter file path to unsigned in index map
+std::unordered_map<std::string, unsigned int> idx_cspline;
+
+// Knots for each curve
+std::vector< std::vector<double> > knots_cspline;
+
+// Cubic segment parameters for each curve
+std::vector< std::vector<double> > a1_cspline;
+std::vector< std::vector<double> > a2_cspline;
+std::vector< std::vector<double> > a3_cspline;
+std::vector< std::vector<double> > a4_cspline;
+
+// Number of segments in each curve
+std::vector< unsigned int > n_seg_cspline;
+
+// The data file format is a text file with one parameter per line
+//   Line 1: Number of parameters
+//   nsegs + 1 lines for knots
+//   nsegs lines for a1
+//   nsegs lines for a2
+//   nsegs lines for a3
+//   nsegs lines for a4 
+
+// Function to read parameter file if not read and store params
+static int read_parameters_cspline(std::string file_path) {
+   try{ //if read already return curve index and done.
+      return idx_cspline.at(file_path);
+   }
+   catch(std::out_of_range const&){} // Not read so read it.
+   unsigned int i, idx, n; //loop counter, curve index, and number of segs
+   std::ifstream param_file; // parameter input file
+
+   // Set index for cspline param file and increment curve count
+   idx_cspline[file_path] = n_cspline;
+   idx = n_cspline;
+   ++n_cspline;
+
+   // open the parameter file for input
+   param_file.open(file_path);
+
+   // get the number of segments and size to vectors
+   param_file >> n;
+   n_seg_cspline.resize(n_cspline);
+   n_seg_cspline[idx] = n;
+   knots_cspline.resize(n_cspline);
+   a1_cspline.resize(n_cspline);
+   a2_cspline.resize(n_cspline);
+   a3_cspline.resize(n_cspline);
+   a4_cspline.resize(n_cspline);
+   knots_cspline[idx].resize(n + 1);
+   a1_cspline[idx].resize(n);
+   a2_cspline[idx].resize(n);
+   a3_cspline[idx].resize(n);
+   a4_cspline[idx].resize(n);
+
+   // get the knots
+   for(i=0; i < n + 1; ++i){
+      param_file >> knots_cspline[idx][i];
+   }
+
+   // get the a1 params
+   for(i=0; i < n; ++i){
+      param_file >> a1_cspline[idx][i];
+   }
+
+   // get the a2 params
+   for(i=0; i < n; ++i){
+      param_file >> a2_cspline[idx][i];
+   }
+
+   // get the a3 params
+   for(i=0; i < n; ++i){
+      param_file >> a3_cspline[idx][i];
+   }
+
+   // get the a4 params
+   for(i=0; i < n; ++i){
+      param_file >> a4_cspline[idx][i];
+   }
+
+   param_file.close();
+
+   // Returns curve index
+   return idx;
+}
+
+// Calculate the cspline 1D function value and derivatives
+extern real cspline(arglist *al) {
+   // Get the data file path argument
+   const char* data_file = al->sa[-(al->at[1]+1)];
```

Comment:
Is there any use / utility to support passing the parameters in directly as arguments and avoiding files?  [probably not, as it would make the Jacobian / Hessian matrices huge?]

Reply:
We could just pass the contents of the parameter file as a string, so we could have an option file name or file contents.  After the first time you call the function with a particular set of parameters, they are cached and not read again.  I think it would definitely be nice not have to write a file, but I wasn't sure how passing a string that large would affect performance.  Regardless, it may be a good idea.  Like you mentioned parameters as numerical arguments would make for a lot of derivatives.

Reply:
I like that idea.  I don't know if the ASL has a string length limit (but quickly looking through the documentation doesn't indicate a limit).

Reply:
Seems to work fine.  I added that option.  The way the file format is setup, there is one number on each line, so I just assume if the string contains a new line, it is contents and if not, it is a file name.  Hopefully there are no cases where valid file names contain newlines.